OpenHAB stops watching any files for changes

This has gotten to the point where I have to restart openhab constantly to see changes to my items or rules files. I'm on openhab 3.2.0 S2517. The symptom is that no file changes are ever noticed - I have to restart openhab to pick up any changes. I enabled debug logging for org.openhab.core.service (and some the ScriptFileService, as I'm using the JRuby add-on and that's where I first noticed it, but the problem also applies to regular .items files), and I see this in my log when I boot up:

```
2021-10-24 18:26:04.268 [INFO ] [.core.model.lsp.internal.ModelServer] - Started Language Server Protocol (LSP) service on port 5007
2021-10-24 18:26:04.322 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'controlpanels.rules'
2021-10-24 18:26:04.734 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'pool.rules'
2021-10-24 18:26:05.247 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'locks.rules'
2021-10-24 18:26:05.630 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'pigpio.rules'
2021-10-24 18:26:05.771 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'water.rules'
2021-10-24 18:26:05.887 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'myq.rules'
2021-10-24 18:26:06.020 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'tv.rules'
2021-10-24 18:26:06.263 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'alarm.rules'
2021-10-24 18:26:07.070 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'usg.rules'
2021-10-24 18:26:07.144 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'minisplit.rules'
2021-10-24 18:26:07.443 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'astro.rules'
2021-10-24 18:26:07.692 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'computers.rules'
2021-10-24 18:26:07.783 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'heatpump.rules'
2021-10-24 18:26:08.054 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'switches.rules'
2021-10-24 18:26:08.641 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'doorbell.rules'
2021-10-24 18:26:08.714 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'abb.rules'
2021-10-24 18:26:09.099 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'shades.rules'
2021-10-24 18:26:09.297 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'bwa.rules'
2021-10-24 18:26:09.371 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'status.rules'
2021-10-24 18:26:10.997 [INFO ] [org.openhab.ui.internal.UIService   ] - Started UI on port 8080
2021-10-24 18:26:15.074 [DEBUG] [rt.internal.loader.ScriptFileWatcher] - Enqueued file:/etc/openhab/automation/jsr223/ruby/personal/switches.rb
2021-10-24 18:26:15.085 [DEBUG] [enhab.core.service.StartLevelService] - Reached start level 20
2021-10-24 18:26:15.085 [DEBUG] [rt.internal.loader.ScriptFileWatcher] - Enqueued file:/etc/openhab/automation/jsr223/ruby/personal/pool.rb
2021-10-24 18:26:15.086 [DEBUG] [rt.internal.loader.ScriptFileWatcher] - Enqueued file:/etc/openhab/automation/jsr223/ruby/personal/hiome.rb
2021-10-24 18:26:15.086 [DEBUG] [rt.internal.loader.ScriptFileWatcher] - Enqueued file:/etc/openhab/automation/jsr223/ruby/personal/keypads.rb
2021-10-24 18:26:15.087 [DEBUG] [rt.internal.loader.ScriptFileWatcher] - Enqueued file:/etc/openhab/automation/jsr223/ruby/personal/air_quality.rb
2021-10-24 18:26:15.087 [DEBUG] [rt.internal.loader.ScriptFileWatcher] - Enqueued file:/etc/openhab/automation/jsr223/ruby/personal/bond.rb
2021-10-24 18:26:15.087 [DEBUG] [rt.internal.loader.ScriptFileWatcher] - Enqueued file:/etc/openhab/automation/jsr223/ruby/personal/test.rb
2021-10-24 18:26:15.088 [DEBUG] [rt.internal.loader.ScriptFileWatcher] - Enqueued file:/etc/openhab/automation/jsr223/ruby/personal/scenes.rb
2021-10-24 18:26:15.088 [DEBUG] [rt.internal.loader.ScriptFileWatcher] - Enqueued file:/etc/openhab/automation/jsr223/ruby/personal/shades.rb
2021-10-24 18:26:15.089 [DEBUG] [rt.internal.loader.ScriptFileWatcher] - Enqueued file:/etc/openhab/automation/jsr223/ruby/personal/everlights.rb
2021-10-24 18:26:15.958 [DEBUG] [penhab.core.service.WatchQueueReader] - ClosedWatchServiceException caught! null. 
Dir Watcher Stopping
2021-10-24 18:26:18.163 [INFO ] [.transport.mqtt.MqttBrokerConnection] - Starting MQTT broker connection to <redacted>
2021-10-24 18:26:18.209 [INFO ] [.transport.mqtt.MqttBrokerConnection] - Starting MQTT broker connection to <redacted>
2021-10-24 18:26:18.422 [ERROR] [al.handler.DSCAlarmBaseBridgeHandler] - Not Connected to the DSC Alarm!
2021-10-24 18:26:18.741 [INFO ] [.onkyo.internal.handler.OnkyoHandler] - Using configuration: <redacted>
2021-10-24 18:26:18.748 [INFO ] [.onkyo.internal.handler.OnkyoHandler] - Using configuration: <redacted>
2021-10-24 18:26:18.752 [INFO ] [.onkyo.internal.handler.OnkyoHandler] - Using configuration: <redacted>
2021-10-24 18:26:18.896 [INFO ] [zwave.handler.ZWaveControllerHandler] - Attempting to add listener when controller is null
2021-10-24 18:26:18.957 [INFO ] [mekit.internal.HomekitChangeListener] - Created 129 HomeKit items.
2021-10-24 18:26:20.120 [DEBUG] [enhab.core.service.StartLevelService] - Reached start level 30
2021-10-24 18:26:20.121 [DEBUG] [enhab.core.service.StartLevelService] - Reached start level 40
```

It seems the WatchQueueReader run thread (https://github.com/openhab/openhab-core/blob/569dddd046d127bbf1f40d64c7687abb70db4d74/bundles/org.openhab.core/src/main/java/org/openhab/core/service/WatchQueueReader.java#L277) is crashing, for reasons unknown. I would guess the enqueuing of the scripts, but all of my scripts in the directory have been `Enqueued` by that point, and enqueuing doesn't look to do anything directly in that thread.

I'd be happy to enable additional debugging or run a development build to get more details.

@boc-tothefuture I tried this PR. It seems to fix our problem with jruby bundle seting GEM_HOME to conf/automation/lib/ even though it did raise a little exception, the file watcher continued to work after that.

```
2021-11-08 15:28:39.346 [INFO ] [org.openhab.ui.internal.UIService   ] - Started UI on port 8484
2021-11-08 15:28:44.087 [INFO ] [e.automation.internal.RuleEngineImpl] - Rule engine started.
2021-11-08 15:28:48.319 [ERROR] [penhab.core.service.WatchQueueReader] - Exception caught in WatchQueueReader:
java.lang.NullPointerException: null
	at org.openhab.core.service.WatchQueueReader.resolvePath(WatchQueueReader.java:338) ~[bundleFile:?]
	at org.openhab.core.service.WatchQueueReader.run(WatchQueueReader.java:225) [bundleFile:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
2021-11-08 15:28:48.336 [ERROR] [penhab.core.service.WatchQueueReader] - Exception caught in WatchQueueReader:
java.lang.NullPointerException: null
	at org.openhab.core.service.WatchQueueReader.resolvePath(WatchQueueReader.java:338) ~[bundleFile:?]
	at org.openhab.core.service.WatchQueueReader.run(WatchQueueReader.java:225) [bundleFile:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
2021-11-08 15:28:48.930 [TRACE] [penhab.core.service.WatchQueueReader] - Modification event for /openhab/conf/automation/lib/ruby/gem_home/specifications/openhab-scripting-4.13.3.gemspec
2021-11-08 15:28:48.931 [TRACE] [penhab.core.service.WatchQueueReader] - Scheduled processing of /openhab/conf/automation/lib/ruby/gem_home/specifications/openhab-scripting-4.13.3.gemspec
2021-11-08 15:28:49.963 [TRACE] [penhab.core.service.WatchQueueReader] - Executing job for /openhab/conf/automation/lib/ruby/gem_home/specifications/openhab-scripting-4.13.3.gemspec
2021-11-08 15:28:49.963 [TRACE] [penhab.core.service.WatchQueueReader] - Job removed itself for /openhab/conf/automation/lib/ruby/gem_home/specifications/openhab-scripting-4.13.3.gemspec
2021-11-08 15:28:49.969 [WARN ] [ty.util.ssl.SslContextFactory.config] - Trusting all certificates configured for Client@4974c556[provider=null,keyStore=null,trustStore=null]
2021-11-08 15:28:49.970 [WARN ] [ty.util.ssl.SslContextFactory.config] - No Client EndPointIdentificationAlgorithm configured for Client@4974c556[provider=null,keyStore=null,trustStore=null]
2021-11-08 15:28:50.344 [WARN ] [ty.util.ssl.SslContextFactory.config] - Trusting all certificates configured for Client@7f78d7ba[provider=null,keyStore=null,trustStore=null]
2021-11-08 15:28:50.361 [WARN ] [ty.util.ssl.SslContextFactory.config] - No Client EndPointIdentificationAlgorithm configured for Client@7f78d7ba[provider=null,keyStore=null,trustStore=null]
```

--- a/bundles/org.openhab.core/src/main/java/org/openhab/core/service/WatchQueueReader.java
+++ b/bundles/org.openhab.core/src/main/java/org/openhab/core/service/WatchQueueReader.java
@@ -337,8 +337,11 @@ public class WatchQueueReader implements Runnable {
         Path baseWatchedDir = null;
         Path registeredPath = null;
         synchronized (this) {
-            baseWatchedDir = keyToService.get(key).getSourcePath();
-            registeredPath = registeredKeys.get(key);
+            AbstractWatchService service = keyToService.get(key);
+            if (service != null) {
+                baseWatchedDir = service.getSourcePath();
+                registeredPath = registeredKeys.get(key);
+            }
         }
         if (registeredPath != null) {
             // If the path has been registered in the watch service it relative path can be resolved

